### ISSUE TITLE
Accuracy score from DeDe and computed using sklearn

### ISSUE THREAD

**Issue by EBazarov** [beniz/deepdetect#170]
I have a difficulty of understanding metrics from DeDe. When I'm providing Train and Test data set, DeDe compute metrics during training, and at the end provide final results. But when I predict same test set, using "predict API" and then compute accuracy metric, I had results that are not the same at all. I'm using SVM connector, and for testing purpose decided to use public dataset _agnews_. 

You can find my code and results in attached html report witch extracted from jupyter notebook.
Maybe it's not problem of metrics, but problem of wrong using DeDe API, I'm confusing, because  difference is huge.

[DeDe_SVM_agnews_test_acc.zip](https://github.com/beniz/deepdetect/files/455607/DeDe_SVM_agnews_test_acc.zip)

**Comment by beniz** [comment]
I believe your python code reproduced below is not doing what I suppose you want it to do:

```python
def read_dede_results(dic_results):
    pred_1_prob = []
    for pred in dic_results['body']['predictions']:
        pred_1_prob.append(pred['classes'][1]['prob'])
    y_pred_proba = np.array(pred_1_prob).astype(np.float)
    y_pred = np.around(np.array(pred_1_prob))
    return y_pred_proba, y_pred
```

- the top category and probability should be acquired from `pred['classes'][0]`, not sure why you are using the second value instead;

- the label should be acquired from `pred['classes'][0]['cat']`, instead you are rounding all probabilities to 0

In the end my understanding is that you are computing the accuracy of a classifier that returns label 0 everywhere.

**Comment by EBazarov** [comment]
I think I get you, you mean that predicted classes sorted in results, that's mean you have at first place best label.

Sorry my mistake, I will rerun the code. But the problem that before I have used API with 'best':1, and extracted predictions in a little bit more intelligent way, but also had differents. 

Thanks, I will do one m

**Comment by EBazarov** [comment]
I will do test one more time ASAP

**Comment by EBazarov** [comment]
Ok, this is a new results. I have extracted prediction using **pred['classes'][0]['cat']**, so now I think it should be rigth.

Still have difference between accuracy from DeDe and computed using sklearn.

[DeDe_SVM_agnews_test_acc_bug_fix.zip](https://github.com/beniz/deepdetect/files/456096/DeDe_SVM_agnews_test_acc_bug_fix.zip)

**Comment by beniz** [comment]
AGnews has 4 balanced classes, you are only declaring two to dd, with `nclasses=2`. You can easily observe this since the predictions are then only 0 or 1. The probability that a 4 balanced classes classifier with 90% accuracy outputs only 0 and 1 is null. From there you can immediately infer that you have a class specification problem...

**Comment by EBazarov** [comment]
At top of a notebook file I have declared that I will use only "Sport" and "Business" data for simplicity.

So that's why I am using nclasses = 2

**Comment by beniz** [comment]
It is even worse then, it is impossible to have a binary classifier do 38% accuracy, since reversing the classes would yield 62% accuracy...

**Comment by beniz** [comment]
OK, now I can reproduce through another mean, thanks.

**Comment by ArdalanM** [comment]
I am able to reproduce this behavior with the following code

accuracy calculated by DD: ~90%
accuracy calculated from DD probabilities: ~64%

```python
import time
import numpy as np
from sklearn.datasets import load_svmlight_file
from sklearn.metrics import accuracy_score
from dd_client import DD

sname = "AccuracyComparison"
description = "Reproduce accuracy measure"

# change to your model path
repository = "/path/to/model"

# path to data (binary classification data)
input_data = ["X_train.svm", "X_test.svm"]


# Service creation
dd = DD('localhost')
dd.set_return_format(dd.RETURN_PYTHON)
model = {'templates': '../templates/caffe', 'repository': repository}
mllib = 'caffe'

service_parameters_input = {'connector': 'svm'}
service_parameters_mllib = {"db": True, "template": "mlp", "nclasses": 2, "layers": [128,128], "activation": 'relu'}
service_parameters_output = {}

json_dump = dd.put_service(sname, model, description, mllib, service_parameters_input,
                                service_parameters_mllib, service_parameters_output)


time.sleep(1)


# Training model
train_parameters_input = {'db': True},
train_parameters_output = {"measure": ['mcll','auc','accp']},
train_parameters_mllib = {'gpu': True,
                               'solver': {'iterations': 1000, 'test_interval': 100, 'base_lr': 0.01,
                                          'solver_type': 'adam'},
                               'net': {'batch_size': 256}}

dd.post_train(sname, input_data,
              train_parameters_input,
              train_parameters_mllib,
              train_parameters_output, async=True)

train_status = ''
while True:
    train_status = dd.get_train(sname, job=1, timeout=2)
    if train_status['head']['status'] == 'running':
        print(train_status['body']['measure'])
    else:
        print(train_status)
        break



# Predicting test set
predict_parameters_input = {},
predict_parameters_output = {"best": 2},
predict_parameters_mllib = {'gpu': True, 'gpuid': 0}

json_dump = dd.post_predict(sname, [input_data[1]], predict_parameters_input, predict_parameters_mllib,
                            predict_parameters_output)


# From json_dump to array
nb_rows = len(json_dump['body']['predictions'])
nb_col = 2
y_score = np.zeros((nb_rows, nb_col))
for i, row in enumerate(json_dump['body']['predictions']):
    row_number = int(row['uri'])
    assert row_number == i

    for classe in row['classes']:
        class_label = int(classe['cat'])
        class_prob = classe['prob']
        y_score[row_number, class_label] = class_prob
assert np.isclose(y_score[:, 0] + y_score[:, 1], np.ones((len(y_score)))).all()


# Loading labels from test file
X, Y = load_svmlight_file(input_data[1])
acc = accuracy_score(Y, y_score.argmax(-1))
print("Accuracy calculed from DD probabilities: {}".format(acc))
# dd.delete_service(sname,"full")
```

**Comment by beniz** [comment]
Thanks, you can actually reproduce by simply adding `measure:['f1']` to the predict call in the `output` object.

**Comment by beniz** [comment]
Fixed. Some would appreciate the hellish effect of a misplaced `-1`... So using `/predict` along with `measure` allows you to get the accuracy on test set whenever that is needed outside of the training loop.
